cooked-validators/src/Cooked/MockChain/BlockChain.hs

```diff
@@ -278,6 +277,83 @@ valueFromTxOutRef oref = do
 
+-- ** Slot and Time Management
+
+-- $slotandtime
+-- #slotandtime#
+--
+-- Slots are integers that monotonically increase and model the passage of time. By looking
+-- at the current slot, a validator gets to know that it is being executed within a certain
+-- window of wall-clock time. Things can get annoying pretty fast when trying to mock traces
+-- and trying to exercise certain branches of certain validators; make sure you also read
+-- the docs on 'autoSlotIncrease' to be able to simulate sending transactions in parallel.
+
+-- | Moves n slots fowards
+waitNSlots :: (MonadBlockChainWithoutValidation m) => Integer -> m Ledger.Slot
+waitNSlots n =
+  if n < 0
+    then fail "waitNSlots: negative argument"
+    else currentSlot >>= awaitSlot . (+ fromIntegral n)
+
+-- | Returns the closed ms interval corresponding to the current slot
+currentTime :: (MonadBlockChainWithoutValidation m) => m (PV2.POSIXTime, PV2.POSIXTime)
+currentTime = slotToTimeInterval =<< currentSlot
+
+-- | Returns the closed ms interval corresponding to the slot with the given
+-- number. It holds that
+--
+-- > slotToTimeInterval (getEnclosingSlot t) == (a, b)    ==>   a <= t <= b
+--
+-- and
+--
+-- > slotToTimeInterval n == (a, b)   ==>   getEnclosingSlot a == n && getEnclosingSlot b == n
+--
+-- and
+--
+-- > slotToTimeInterval n == (a, b)   ==>   getEnclosingSlot (a-1) == n-1 && getEnclosingSlot (b+1) == n+1
+slotToTimeInterval :: (MonadBlockChainWithoutValidation m) => Ledger.Slot -> m (PV2.POSIXTime, PV2.POSIXTime)
+slotToTimeInterval slot = do
+  slotConfig <- Emulator.pSlotConfig <$> getParams
+  case Emulator.slotToPOSIXTimeRange slotConfig slot of
+    PV2.Interval
+      (PV2.LowerBound (PV2.Finite l) leftclosed)
+      (PV2.UpperBound (PV2.Finite r) rightclosed) ->
+        return
+          ( if leftclosed then l else l + 1,
+            if rightclosed then r else r - 1
+          )
+    _ -> error "The time interval corresponding to a slot should be finite on both ends."
+
+-- | Return the slot that contains the given time. See 'slotToTimeInterval' for
+-- some equational properties this function satisfies.
+getEnclosingSlot :: (MonadBlockChainWithoutValidation m) => PV2.POSIXTime -> m Ledger.Slot
+getEnclosingSlot t = do
+  slotConfig <- Emulator.pSlotConfig <$> getParams
+  return $ Emulator.posixTimeToEnclosingSlot slotConfig t
+
+-- | Waits until the current slot becomes greater or equal to the slot containing the given POSIX time.
+--  Note that that it might not wait for anything if the current slot is large enough.
+awaitEnclosingSlot :: (MonadBlockChainWithoutValidation m) => PV2.POSIXTime -> m Ledger.Slot
+awaitEnclosingSlot = awaitSlot <=< getEnclosingSlot
+
+-- | The infinite range of slots ending before or at the given POSIX time
+slotRangeBefore :: MonadBlockChain m => PV2.POSIXTime -> m Ledger.SlotRange
+slotRangeBefore t = do
+  n <- getEnclosingSlot t
+  (_, b) <- slotToTimeInterval n
+  if t == b
```

Comment:
Took me a bit to convince myself the equality check is good here (instead of inequality check). Might be worth adding a comment.